### ISSUE TITLE
About the SFT training scripts

### ISSUE THREAD

**Issue by Columbine21** [Fr0zenCrane/UniCoT#18]
Dear Authors, thanks for the great work, However, I can not find the training script, or how to start the sft training from the current codebase.

**Comment by Fr0zenCrane** [comment]
Hi @Columbine21, thank you for your notification and interest in Uni-CoT! 

I’ve updated the README’s Training section with guidance: https://github.com/Fr0zenCrane/UniCoT?tab=readme-ov-file#training. The training codebase of Uni-CoT is adopted from Bagel, their training guidance described in [TRAIN.md](https://github.com/ByteDance-Seed/Bagel/blob/main/TRAIN.md) is also very helpful. 
Please take a look of these two documents and feel free to reach out with any follow-up questions!

**Comment by Columbine21** [comment]
@Fr0zenCrane Thanks for your reply. What about the training data? Are the training data open sourced?

**Comment by Fr0zenCrane** [comment]
You may refer to these two datasets:
UniCoT-Self-Reflection-6K: https://huggingface.co/datasets/Fr0zencr4nE/UniCoT-Self-Reflection-6K
UniCoT-Breakdown-3K: https://huggingface.co/datasets/Fr0zencr4nE/UniCoT-Breakdown-3K

**Comment by Hanbo-Cheng** [comment]
@Fr0zenCrane 
Hello~ Could you please clarify whether these data can be directly used with the Dataset class provided by BAGEL, or if I need to perform additional data format processing, or rewrite a new Dataset class?

**Comment by Fr0zenCrane** [comment]
@Hanbo-Cheng 
For these two datasets, you need to format the data following [TRAIN.md](https://github.com/ByteDance-Seed/Bagel/blob/main/TRAIN.md) adopted from Bagel and the dataset class provided by us: [mdp_dataset.py](https://github.com/Fr0zenCrane/UniCoT/blob/main/data/interleave_datasets/mdp_dataset.py)

**Comment by Hanbo-Cheng** [comment]
@Fr0zenCrane Sorry, I'm still a bit confused. The multi-turn edit [dataset](https://huggingface.co/datasets/AILab-CVC/SEED-Data-Edit-Part2-3/tree/main/multi_turn_editing)  used in Bagel has different format from the two datasets you provided. I'm not sure which dataset class in `data/interleave_datasets/mdp_dataset.py` should be used for loading, nor am I clear about the specific parquet format it should be converted to. Would it be convenient for you to provide some additional hints? It would be even better if you could share the corresponding data processing script.

**Comment by Fr0zenCrane** [comment]
We formulate the interleaved reasoning process as a Markov chain. Consequently, the model's abilities related to image understanding, generation, and editing are all involved.

Taking Self-Reflection as an example:
-  The model will first generate an image conditioned on the input prompt, then evaluate its generated image and refine the image if needed. 
- The evaluation process takes in the generated image and output evaluation on the generated image and give refinement suggestion. 
- At last, the model edits the generated image according to the refinement suggestions.
- This eval-and-then-refine procedure will incurs multiple turns until the model finally think its refined image is flawless. 

Formally, the process can be represented as:
- [generation] → { [evaluation] → [refinement] } × n → [evaluation: everything is good]

Each of [generation], [evaluation], and [refinement] requires a dataset class. You can use the image generation, editing, and understanding dataset classes provided by Bagel, or combine Bagel’s understanding dataset with our image generation and editing datasets. Both configurations are tested and perform well. The primary difference is that our setup adds a system prompt to better preserve Bagel’s original capabilities.

While the explanation above covers the core flow and design choices, it may not address every aspect of your inquiry in full detail. You can leave an online meeting link, I suppose 15-30min is enough, better in afternoon or night (UTC+8).

**Comment by Columbine21** [comment]
@Fr0zenCrane 感谢回复，所以您的意思是实际上Bagel CoT的SFT 被拆分成很多子任务的训练？比如从用户prompt 生成 think 文本这一过程为一个 [system prompt] + [user prompt] +[think text] (只监督这部分) 的子任务？ 然后根据 [system prompt] + [user prompt] + [think text] (只监督这部分)  生成图像 也被单独定义为一个任务, 根据 [system prompt] + [生成图像] 生成 [Evaluation] (监督) 被定义为一个子任务？其他的部分也类似，是不是可以这么理解？有点没理解代码中

<img width="749" height="492" alt="Image" src="https://github.com/user-attachments/assets/b4965835-0314-4f8e-885f-495172e6ab73" />

这个step的划分分别是想训练什么子任务。感谢您的回复

**Comment by Fr0zenCrane** [comment]
@Columbine21 对的对的，我们将一类马尔可夫状态转移过程（一个子任务）对应到训练用到的一个Dataset Class上。

最后这个数据集是为了训练这类科学背景的生成任务：[Nano-Banana-Style Geography Reasoning Generation](https://github.com/Fr0zenCrane/UniCoT/blob/main/assets/qualitative_results_geo.png)。

我们发现直接去做这些科学任务，由于gap比较大，而且这些任务比较抽象和困难，直接做是做不好的，所以我们将他拆分成了4个子任务，每次生成都要依次进行这4个子任务（走一个长度为4的马尔可夫过程），得到最终的图片。这样，在代码实现的时候，这4个step，每一个step对应一个数据集。但我实现的时候发现真这么搞比较麻烦。因为Bagel的数据集采样逻辑是按照dataset group来管理的，一个dataset group必须是同样的dataset class，这样，为了一个科学任务我就需要新增4个dataset group，而且不利于平衡科学数据和普通数据的比例。有鉴于此，我用这种写法强行将4种不同的子任务耦合在一个dataset class，方便数据集采样权重平衡

**Comment by WangYijun-OUC** [comment]
作者你好！想问下在[mdp_dataset.py](https://github.com/Fr0zenCrane/UniCoT/blob/main/data/interleave_datasets/mdp_dataset.py)中定义的MDPedit和MDPT2I等数据类是在RL代码中才会用到的吗？
另外，关于SFT阶段，是否可以给出你们在t2i和edit任务中所使用的parquet数据集或者转换的脚本？我正在使用你给出的reflection-6K数据，但实在不知道如何将其与data_dir或parquet_info_path对应起来。感谢！
